core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -998,7 +1042,13 @@ protected void innerOnFailure(Exception e) {
             logger.warn(
                 (Supplier<?>) () -> new ParameterizedMessage("exception caught on transport layer [{}], closing connection", channel), e);
             // close the channel, which will cause a node to be disconnected if relevant
-            closeChannelWhileHandlingExceptions(channel);
+            TcpChannelUtils.closeChannel(channel, false);
+        }
+    }
+
+    protected void serverAcceptedChannel(Channel channel) {
+        if (acceptedChannels.add(channel)) {
```

Comment:
I am not sure why we have this conditional. Do we deduplicate? do we have duplicates? I think we should not do this but rather make sure we assert on it or throw a hard exception? I also wonder if we should install the close listener first otherwise it might not be invoked in the case of a race?

Reply:
I agree that the channel should not have be attempted to be added twice. So I added an assertion. 

I actually disagree on the other point. The only way to avoid a race is to add the close listener afterwards. Imagine that the channel is already closed. We add a close listener that removes the channel and it is immediately executed. Then we add the channel to the accepted set. Nothing removes it at this point. 

Close listeners are always invoked once they are added. Even if they are added after a channel is closed (they are invoked immediately). So it is safe to add the close listener after the channel is added to the accepted set, because it will ensure that the channel is actually removed.